### ISSUE TITLE
Configurable trim type

### ISSUE THREAD

**Issue by MarkMLl** [inkstitch/inkstitch#946]
"Auto-Route Satin Columns" needs to be able to output different commands in response to "Trim jump stitches" depending on the capability of the output format and machine.

A Brother Innovis-97E ignores trim commands and appears to crash if it loads a .pes containing a pause, the only way to allow manual intervention is to insert a colour change which is what's in normal embroidery output.

As an experienced developer (albeit no lover of Python) this is probably something I could fix myself, at least for a proof of concept, but I can't find any discussion on how to debug and modify the Inkstitch/pyembroidery combination.

**Comment by kaalleen** [comment]
We do not have documentation about it, sorry. Pyembroidery in Ink/Stitch is basically in action during input and output.
We use a slightly diverged fork from @tatarize : https://github.com/EmbroidePy/pyembroidery
He knows a lot about embroidery file formats. He has done  some fixes and additions that would be nice to have in our pyembroidery version. We haven't come to update it yet. I don't know if I can be of help in this matter. If you have any questions I can try to help. ... thanks for your offer to do some work on the code! It's really appreciated.

**Comment by MarkMLl** [comment]
I can see how Inkstitch hangs together, but how does one debug it? As a starting point, how does one decouple the Python and Electron subsystems so one can isolate e.g. the dire delay-on-startup issue?

Is there a test framework which allows the Pyembroidery transformations to be single-stepped etc.?

I think that my PES/PEC dump program is useful since it's a separate codebase, and it is easily enhanceable to edit the file... early efforts have indicated how intolerant the Brother firmware is of e.g. a mismatch between the number of colour-change commands and the number of defined colours.

**Comment by kaalleen** [comment]
@lexelby could you help us out on this one?

**Comment by lexelby** [comment]
I debug using PyCharm, although any pydevd-based debugger will work.  Create a file named DEBUG at the top of your inkstitch clone and then start your debugger server (i.e. start debug mode in pycharm or eclipse).  It will hot a breakpoint early on, at which point you can step through or hit continue and wait for your own breakpoint.  I think there are docs on this somewhere but I can't remember where... Maybe in inkstitch.py?

I wasn't aware that some innovis machines actually crash on trims.  That's troubling!  I ultimately want to add a global preferences system for Ink/Stitch so that you can set things like "convert trims to color changes" and the like.  Currently there's no easy way to do that.  Probably your easiest path forward is to use bin/pyembroidery-convert in pyembroidery to convert to CSV, text search/replace, and convert back to PES.

I'm afraid my availability continues to be quite minimal so I can't help with the code right now  :(

**Comment by MarkMLl** [comment]
@lexelby My admiration for Python knows no bounds... if it were a woman it would be burnt at the stake. But on occasion one has to deal with the devil... any further info you can point me at would be appreciated, e.g. do I have to have a special Inkstitch build.

I've discussed this with Brother support and there's no firmware patches etc. The Innovis-97E ignores (doesn't crash on) trims, but that makes them less than useful as insertions when working on satin conversion.

The machine gives the impression of crashing if asked to load a file with pauses in it, i.e goes dead for a moment and then reboots which I think (wearing my embedded systems hat) is the result of a watchdog timer.

It also appears to do the same if a file is patched to replace spurious colour-change commands (i.e. between fill and stroke of an object, which is documented as not being supposed to happen) with a dummy jump which will of course leave the number of colour changes and the number of declared colours (i.e. thumbnails etc.) out of step.

What I'm probably going to do is investigate the last of those problems first, since my PES dump program can easily be modified to patch out chunks; that will mean that I can at least report on whether that particular problem is fixable by a postprocessing hack without having to climb too far up the Python curve and should be able improve workflow at this end.

The thing that's really dire though is the 70-sec startup. I might be able to investigate that by putting Inkscape etc. into a container, but it would be really useful to be able to attribute that to either Electron or to the Python stuff.

**Comment by lexelby** [comment]
> My admiration for Python knows no bounds... if it were a woman it would be burnt at the stake.

Uhhh... Let's maybe not make casual reference to parts of history where members of this community were murdered.  

> any further info you can point me at would be appreciated, e.g. do I have to have a special Inkstitch build.

Please forgive my brevity, I'm on my phone at the moment.  On inkstitch.org, there's a doc entitled "manual installation", that shows you how to run Ink/Stitch directly out of a clone of the repo.  That's what you want for development and debugging.

**Comment by MarkMLl** [comment]
I could do with a bit of help getting going with the debugging on this. With reference to https://inkstitch.org/developers/inkstitch/manual-setup/

 cd /usr/local/src
 sudo mkdir inkstitch
 // Make directory accessible as necessary
 git clone https://github.com/inkstitch/inkstitch
 // Should this next be preceded by cd ./inkstitch?
 git clone https://github.com/inkstitch/pyembroidery.git

Now at this point I start getting bogged down in the differences between Python2/3. I see reference to Ubuntu 16.04 (rough contemporary is Debian "Lenny") and looking at the Windows instructions I see that Python 2.7 is preferred. Lenny has 2.5, Debian "Stretch" has 2.7.13 so if I'm to stick with Debian then Stretch would appear to be a good starting point since it is on the cusp before the switch to Python3. Any comment on that would be appreciated before I go much further.

I see reference to pyenv in the instructions: how is this used in this context?

I got bogged down yesterday in inappropriate NodeJS versions etc., but that was trying to use a much more recent Debian version.

Noting the comment about PyCharm as a debugging environment, how is Inkscape told to transfer control to this (or am I getting vastly ahead of myself)?

**Comment by lexelby** [comment]
I'm pretty sure our PES trim support is a little broken and there's another issue open about that.

Regarding debugging, sorry for dropping this!  Let me know if you're still interested in debugging Ink/Stitch

**Comment by MarkMLl** [comment]
Definitely not this week and probably not this year. I've got an enormous amount of extraneous crap occupying my time which has hardly helped my business activities, and once I've managed to clear the decks a bit those obviously take priority.

However in the larger scale of things I'm interested in at least getting the project to the point where there's a documented and occasionally-tested debugging methodology.
